langgraph/src/prebuilt/react_agent_executor.ts

```diff
@@ -0,0 +1,187 @@
+import { BaseCheckpointSaver } from "../checkpoint/base.js";
+import { END, START, StateGraph } from "../graph/index.js";
+import { MessagesState } from "../graph/message.js";
+import { CompiledStateGraph, StateGraphArgs } from "../graph/state.js";
+import { All } from "../pregel/types.js";
+import { ToolNode } from "./tool_node.js";
+
+export interface AgentState {
+  messages: BaseMessage[];
+  // TODO: This won't be set until we
+  // implement managed values in LangGraphJS
+  is_last_step: boolean;
+}
+
+export type N = typeof START | "agent" | "tools";
+
+/**
+ * Creates a StateGraph agent that relies on a chat model utilizing tool calling.
+ * @param model The chat model that can utilize OpenAI-style function calling.
+ * @param tools A list of tools or a ToolNode.
+ * @param messageModifier An optional message modifier to apply to messages before being passed to the LLM.
+ * Can be a SystemMessage, string, function that takes and returns a list of messages, or a Runnable.
+ * @param checkpointSaver An optional checkpoint saver to persist the agent's state.
+ * @param interruptBefore An optional list of node names to interrupt before running.
+ * @param interruptAfter An optional list of node names to interrupt after running.
+ * @returns A compiled agent as a LangChain Runnable.
+ */
+export function createReactAgent(
+  model: BaseChatModel,
+  tools: ToolNode<MessagesState> | StructuredTool[],
+  messageModifier?:
+    | SystemMessage
+    | string
+    | ((messages: BaseMessage[]) => BaseMessage[])
+    | Runnable,
+  checkpointSaver?: BaseCheckpointSaver,
+  interruptBefore?: N[] | All,
+  interruptAfter?: N[] | All
+): CompiledStateGraph<
+  AgentState,
+  Partial<AgentState>,
+  typeof START | "agent" | "tools"
+> {
+  const schema: StateGraphArgs<AgentState>["channels"] = {
+    messages: {
+      value: (left: BaseMessage[], right: BaseMessage[]) => left.concat(right),
+      default: () => [],
+    },
+    is_last_step: {
+      value: (_?: boolean, right?: boolean) => right ?? false,
+      default: () => false,
+    },
+  };
+
+  let toolClasses: (StructuredTool | DynamicTool)[];
+  if (!Array.isArray(tools)) {
+    toolClasses = tools.tools;
+  } else {
+    toolClasses = tools;
+  }
+  if (!("bindTools" in model) || typeof model.bindTools !== "function") {
+    throw new Error(`Model ${model} must define bindTools method.`);
+  }
+  const modelWithTools = model.bindTools(toolClasses);
+  const modelRunnable = _createModelWrapper(modelWithTools, messageModifier);
+
+  const shouldContinue = (state: AgentState) => {
+    const { messages } = state;
+    const lastMessage = messages[messages.length - 1];
+    if (
+      isAIMessage(lastMessage) &&
+      (!lastMessage.tool_calls || lastMessage.tool_calls.length === 0)
+    ) {
+      return END;
+    } else {
+      return "continue";
+    }
+  };
+
+  const callModel = async (state: AgentState) => {
+    const { messages } = state;
+    // TODO: Stream
```

Comment:
If we wanted to support streamEvents, i think i need to actually stream here and concatenate, though i need to check. been a while since ir eviewed the js chat model implementations

Reply:
Because in js we don't auto promote invoke to stream like we do in py?

Reply:
afaict